File: FileSystems/IOFileSystem/src/main/java/org/tobi29/scapes/engine/utils/io/filesystem/io/IOFileUtilImpl.java
package org.tobi29.scapes.engine.utils.io.filesystem.io;

import java8.util.function.Predicate;
import java8.util.stream.Stream;
import org.apache.tika.Tika;
import org.threeten.bp.Instant;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.io.*;
import org.tobi29.scapes.engine.utils.io.filesystem.FilePath;
import org.tobi29.scapes.engine.utils.io.filesystem.FileUtilImpl;
import org.tobi29.scapes.engine.utils.io.filesystem.ReadSource;

import java.io.*;
import java.net.URI;
import java.nio.channels.FileChannel;
import java.nio.channels.ReadableByteChannel;
import java.util.ArrayList;
import java.util.List;
import java.util.zip.ZipFile;

public class IOFileUtilImpl implements FileUtilImpl {
    public static File toFile(FilePath path) {
        if (path instanceof FilePathImpl) {
            return ((FilePathImpl) path).file;
        }
        return new File(path.toUri());
    }

    private static void read(File file,
            IOConsumer<BufferedReadChannelStream> read) throws IOException {
        try (FileChannel channel = channelR(file)) {
            read.accept(new BufferedReadChannelStream(channel));
        }
    }

    private static <R> R readReturn(File file,
            IOFunction<BufferedReadChannelStream, R> read) throws IOException {
        try (FileChannel channel = channelR(file)) {
            return read.apply(new BufferedReadChannelStream(channel));
        }
    }

    private static void write(File file,
            IOConsumer<BufferedWriteChannelStream> write) throws IOException {
        try (FileChannel channel = channelDRW(file)) {
            BufferedWriteChannelStream stream =
                    new BufferedWriteChannelStream(channel);
            write.accept(stream);
            stream.flush();
        }
    }

    private static <R> R writeReturn(File file,
            IOFunction<BufferedWriteChannelStream, R> write)
            throws IOException {
        try (FileChannel channel = channelDRW(file)) {
            BufferedWriteChannelStream stream =
                    new BufferedWriteChannelStream(channel);
            R r = write.apply(stream);
            stream.flush();
            return r;
        }
    }

    public static FilePath path(File file) {
        return new FilePathImpl(file);
    }

    public static ReadSource read(File file) {
        return new ReadSource() {
            @Override
            public boolean exists() {
                return file.exists();
            }

            @Override
            public InputStream readIO() throws IOException {
                return new FileInputStream(file);
            }

            @Override
            public void read(IOConsumer<ReadableByteStream> reader)
                    throws IOException {
                IOFileUtilImpl.read(file, reader::accept);
            }

            @Override
            public ReadableByteChannel channel() throws IOException {
                return channelR(file);
            }

            @Override
            public <R> R readReturn(IOFunction<ReadableByteStream, R> reader)
                    throws IOException {
                return IOFileUtilImpl.readReturn(file, reader::apply);
            }

            @Override
            public String mimeType() throws IOException {
                try (InputStream streamIn = readIO()) {
                    return new Tika().detect(streamIn, file.toString());
                }
            }
        };
    }

    public static void deleteDir(File file) {
        // TODO: Add loop check
        File[] files = file.listFiles();
        if (files != null) {
            for (File child : files) {
                if (child.isFile()) {
                    child.delete();
                } else if (child.isDirectory()) {
                    deleteDir(child);
                }
            }
        }
        file.delete();
    }

    private static FileChannel channelR(File file) throws IOException {
        return new RandomAccessFile(file, "r").getChannel();
    }

    private static FileChannel channelRW(File file) throws IOException {
        return new RandomAccessFile(file, "rw").getChannel();
    }

    private static FileChannel channelDRW(File file) throws IOException {file.delete();
        return new RandomAccessFile(file, "rw").getChannel();
    }

    private static void listRecursive(File file, List<FilePath> list) {
        File[] files = file.listFiles();
        if (files != null) {
            for (File child : files) {
                list.add(path(child));
                if (child.isDirectory()) {
                    listRecursive(child, list);
                }
            }
        }
    }

    @SafeVarargs
    private static void listRecursive(File file, List<FilePath> list,
            Predicate<FilePath>... filters) {
        File[] files = file.listFiles();
        if (files != null) {
            for (File child : files) {
                FilePath filePath = path(child);
                boolean valid = true;
                for (Predicate<FilePath> filter : filters) {
                    if (!filter.test(filePath)) {
                        valid = false;
                        break;
                    }
                }
                if (valid) {
                    list.add(filePath);
                }
                if (child.isDirectory()) {
                    listRecursive(child, list, filters);
                }
            }
        }
    }

    @Override
    public FilePath path(String path) {
        return path(new File(path));
    }

    @Override
    public ReadSource read(FilePath path) {
        return read(toFile(path));
    }

    @Override
    public void read(FilePath path, IOConsumer<BufferedReadChannelStream> read)
            throws IOException {
        read(toFile(path), read);
    }

    @Override
    public <R> R readReturn(FilePath path,
            IOFunction<BufferedReadChannelStream, R> read) throws IOException {
        return readReturn(toFile(path), read);
    }

    @Override
    public void write(FilePath path,
            IOConsumer<BufferedWriteChannelStream> write) throws IOException {
        write(toFile(path), write);
    }

    @Override
    public <R> R writeReturn(FilePath path,
            IOFunction<BufferedWriteChannelStream, R> write)
            throws IOException {
        return writeReturn(toFile(path), write);
    }

    @Override
    public FilePath createDirectories(FilePath path) throws IOException {
        toFile(path).mkdirs();
        return path;
    }

    @Override
    public void delete(FilePath path) throws IOException {
        toFile(path).delete();
    }

    @Override
    public boolean deleteIfExists(FilePath path) throws IOException {
        File file = toFile(path);
        if (file.exists()) {
            file.delete();
            return true;
        }
        return false;
    }

    @Override
    public void deleteDir(FilePath path) throws IOException {
        deleteDir(toFile(path));
    }

    @Override
    public boolean exists(FilePath path) {
        return toFile(path).exists();
    }

    @Override
    public boolean isRegularFile(FilePath path) {
        return toFile(path).isFile();
    }

    @Override
    public boolean isDirectory(FilePath path) {
        return toFile(path).isDirectory();
    }

    @Override
    public boolean isHidden(FilePath path) {
        return toFile(path).isHidden();
    }

    @Override
    public boolean isNotHidden(FilePath path) {
        return !isHidden(path);
    }

    @Override
    public FilePath createTempFile(String prefix, String suffix)
            throws IOException {
        return path(File.createTempFile(prefix, suffix));
    }

    @Override
    public FilePath copy(FilePath source, FilePath target) throws IOException {
        read(toFile(source), input -> {
            write(toFile(target),
                    output -> ProcessStream.process(input, output::put));
        });
        return target;
    }

    @Override
    public void stream(FilePath path, IOConsumer<Stream<FilePath>> consumer)
            throws IOException {
        consumer.accept(Streams.of(list(path)));
    }

    @Override
    public List<FilePath> list(FilePath path) throws IOException {
        List<FilePath> list = new ArrayList<>();
        File[] files = toFile(path).listFiles();
        if (files != null) {
            for (File child : files) {
                list.add(path(child));
            }
        }
        return list;
    }

    @Override
    public final List<FilePath> list(FilePath path,
            Predicate<FilePath>[] filters) {
        List<FilePath> list = new ArrayList<>();
        File[] files = toFile(path).listFiles();
        if (files != null) {
            for (File child : files) {
                FilePath filePath = path(child);
                boolean valid = true;
                for (Predicate<FilePath> filter : filters) {
                    if (!filter.test(filePath)) {
                        valid = false;
                        break;
                    }
                }
                if (valid) {
                    list.add(filePath);
                }
            }
        }
        return list;
    }

    @Override
    public void streamRecursive(FilePath path,
            IOConsumer<Stream<FilePath>> consumer) throws IOException {
        consumer.accept(Streams.of(listRecursive(path)));
    }

    @Override
    public List<FilePath> listRecursive(FilePath path) throws IOException {
        List<FilePath> files = new ArrayList<>();
        listRecursive(toFile(path), files);
        return files;
    }

    @Override
    public final List<FilePath> listRecursive(FilePath path,
            Predicate<FilePath>[] filters) {
        List<FilePath> files = new ArrayList<>();
        listRecursive(toFile(path), files, filters);
        return files;
    }

    @Override
    public void setLastModifiedTime(FilePath path, Instant value)
            throws IOException {
        toFile(path).setLastModified(value.toEpochMilli());
    }

    @Override
    public Instant getLastModifiedTime(FilePath path) throws IOException {
        return Instant.ofEpochMilli(toFile(path).lastModified());
    }

    @Override
    public ZipFile zipFile(FilePath path) throws IOException {
        return new ZipFile(toFile(path));
    }

    @Override
    public <R> R tempChannel(FilePath path, IOFunction<FileChannel, R> consumer)
            throws IOException {
        File file = toFile(path);
        R ret;
        try (FileChannel channel = channelRW(file)) {
            ret = consumer.apply(channel);
        }
        // If this delete fails you are probably using M$ Windows and this
        // implementation is only meant for Android
        file.delete();
        return ret;
    }

    private static final class FilePathImpl implements FilePath {
        private final File file;

        private FilePathImpl(File file) {
            this.file = file;
        }

        @Override
        public int compareTo(FilePath o) {
            return file.compareTo(toFile(o));
        }

        @Override
        public int hashCode() {
            return file.hashCode();
        }

        @Override
        public boolean equals(Object obj) {
            return obj instanceof FilePath &&
                    file.equals(toFile((FilePath) obj));
        }

        @Override
        public String toString() {
            return file.toString();
        }

        @Override
        public URI toUri() {
            return file.toURI();
        }

        @Override
        public FilePath resolve(String other) {
            return path(new File(file, other));
        }

        @Override
        public FilePath resolve(FilePath other) {
            return resolve(other.toString());
        }

        @Override
        public FilePath getFileName() {
            return path(new File(file.getName()));
        }

        @Override
        public FilePath toAbsolutePath() {
            return path(file.getAbsoluteFile());
        }

        @Override
        public ReadSource get(String path) {
            return read(toFile(resolve(path)));
        }
    }
}

File: IOFileSystem/src/main/kotlin/org/tobi29/scapes/engine/utils/io/filesystem/io/internal/IOFileUtilImpl.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.utils.io.filesystem.io.internal

import org.threeten.bp.Instant
import org.tobi29.scapes.engine.utils.io.*
import org.tobi29.scapes.engine.utils.io.filesystem.*
import java.io.File
import java.io.RandomAccessFile
import java.net.URI

internal object IOFileUtilImpl : FileUtilImpl {
    override fun path(path: String) = path(File(path))

    override fun path(file: File): FilePath = FilePathImpl(file)

    override fun <R> read(path: FilePath,
                          read: (ReadableByteStream) -> R): R {
        return read(path.toFile(), read)
    }

    override fun <R> write(path: FilePath,
                           write: (WritableByteStream) -> R): R {
        return write(path.toFile(), write)
    }

    override fun createFile(path: FilePath,
                            vararg attributes: FileAttribute<*>): FilePath {
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(
                    "Attributes are not supported on java.io")
        }
        path.toFile().let { file ->
            if (!file.createNewFile()) {
                throw FileAlreadyExistsException(file)
            }
        }
        return path
    }

    override fun createDirectory(path: FilePath,
                                 vararg attributes: FileAttribute<*>): FilePath {
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(
                    "Attributes are not supported on java.io")
        }
        if (!path.toFile().mkdir()) {
            throw FileSystemException(path,
                    reason = "Failed to create directory")
        }
        return path
    }

    override fun createDirectories(path: FilePath,
                                   vararg attributes: FileAttribute<*>): FilePath {
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(
                    "Attributes are not supported on java.io")
        }
        if (!path.toFile().mkdirs()) {
            throw FileSystemException(path,
                    reason = "Failed to create directories")
        }
        return path
    }

    override fun delete(path: FilePath) {
        path.toFile().delete()
    }

    override fun deleteIfExists(path: FilePath): Boolean {
        val file = path.toFile()
        if (file.exists()) {
            file.delete()
            return true
        }
        return false
    }

    override fun deleteDir(path: FilePath) {
        deleteDir(path.toFile())
    }

    override fun exists(path: FilePath,
                        vararg options: LinkOption): Boolean {
        val nofollow = options.contains(NOFOLLOW_LINKS)
        if (nofollow) {
            throw UnsupportedOperationException(
                    "NOFOLLOW_LINKS is not supported on java.io")
        }
        return path.toFile().exists()
    }

    override fun isRegularFile(path: FilePath,
                               vararg options: LinkOption): Boolean {
        val nofollow = options.contains(NOFOLLOW_LINKS)
        if (nofollow) {
            throw UnsupportedOperationException(
                    "NOFOLLOW_LINKS is not supported on java.io")
        }
        return path.toFile().isFile
    }

    override fun isDirectory(path: FilePath,
                             vararg options: LinkOption): Boolean {
        val nofollow = options.contains(NOFOLLOW_LINKS)
        if (nofollow) {
            throw UnsupportedOperationException(
                    "NOFOLLOW_LINKS is not supported on java.io")
        }
        return path.toFile().isDirectory
    }

    override fun isHidden(path: FilePath): Boolean {
        return path.toFile().isHidden
    }

    override fun isNotHidden(path: FilePath): Boolean {
        return !isHidden(path)
    }

    override fun createTempFile(prefix: String,
                                suffix: String,
                                vararg attributes: FileAttribute<*>): FilePath {
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(
                    "Attributes are not supported on java.io")
        }
        return path(kotlin.io.createTempFile(prefix, suffix))
    }

    override fun createTempDir(prefix: String,
                               vararg attributes: FileAttribute<*>): FilePath {
        if (attributes.isNotEmpty()) {
            throw UnsupportedOperationException(
                    "Attributes are not supported on java.io")
        }
        return path(kotlin.io.createTempDir(prefix))
    }

    override fun copy(source: FilePath,
                      target: FilePath): FilePath {
        read(source.toFile(), { input ->
            write(target.toFile(),
                    { output ->
                        process(input, { output.put(it) })
                    })
        })
        return target
    }

    override fun move(source: FilePath,
                      target: FilePath): FilePath {
        source.toFile().renameTo(target.toFile())
        return target
    }

    override fun <R> list(path: FilePath,
                          consumer: (Sequence<FilePath>) -> R): R {
        return consumer(list(path).asSequence())
    }

    override fun list(path: FilePath): List<FilePath> {
        return path.toFile().listFiles()?.map { path(it) } ?: emptyList()
    }

    override fun <R> listRecursive(path: FilePath,
                                   consumer: (Sequence<FilePath>) -> R): R {
        return consumer(listRecursive(path).asSequence())
    }

    override fun listRecursive(path: FilePath): List<FilePath> {
        val files = ArrayList<FilePath>()
        listRecursive(path.toFile(), files)
        return files
    }

    override fun setLastModifiedTime(path: FilePath,
                                     value: Instant) {
        path.toFile().setLastModified(value.toEpochMilli())
    }

    override fun getLastModifiedTime(path: FilePath): Instant {
        return Instant.ofEpochMilli(path.toFile().lastModified())
    }

    override fun <R> tempChannel(path: FilePath,
                                 consumer: (FileChannel) -> R): R {
        val file = path.toFile()
        return try {
            channelRW(file).use { consumer(it) }
        } finally {
            file.delete()
        }
    }

    private data class FilePathImpl(val file: File) : FilePath {
        override fun compareTo(other: FilePath): Int {
            return file.compareTo(other.toFile())
        }

        override fun toString(): String {
            return file.toString()
        }

        override fun toUri(): URI {
            return file.toURI()
        }

        override fun toFile(): File = file

        override fun normalize(): FilePath {
            return path(file.normalize())
        }

        override fun resolve(other: String): FilePath {
            return path(File(file, other))
        }

        override fun resolve(other: FilePath): FilePath {
            return resolve(other.toString())
        }

        override fun startsWith(other: String): Boolean {
            return file.startsWith(other)
        }

        override fun startsWith(other: FilePath): Boolean {
            return file.startsWith(other.toFile())
        }

        override fun relativize(other: FilePath): FilePath? {
            return file.relativeToOrNull(other.toFile())?.let { path(it) }
        }

        override val fileName get() = path(File(file.name))

        override fun toAbsolutePath(): FilePath {
            return path(file.absoluteFile)
        }

        override fun exists(): Boolean {
            return file.exists()
        }

        override fun channel(): ReadableByteChannel {
            return channelR(file)
        }

        override fun <R> read(reader: (ReadableByteStream) -> R): R {
            return read(file, reader)
        }

        override fun mimeType(): String {
            return read { detectMime(it, file.toString()) }
        }
    }

    private fun <R> read(file: File,
                         read: (ReadableByteStream) -> R): R {
        channelR(file).use { channel ->
            return read(BufferedReadChannelStream(channel))
        }
    }

    private fun <R> write(file: File,
                          write: (WritableByteStream) -> R): R {
        channelDRW(file).use { channel ->
            val stream = BufferedWriteChannelStream(channel)
            val r = write(stream)
            stream.flush()
            return r
        }
    }

    private fun deleteDir(file: File) {
        // TODO: Add loop check
        val files = file.listFiles()
        if (files != null) {
            for (child in files) {
                if (child.isFile) {
                    child.delete()
                } else if (child.isDirectory) {
                    deleteDir(child)
                }
            }
        }
        file.delete()
    }

    private fun channelR(file: File): FileChannel {
        return RandomAccessFile(file, "r").channel
    }

    private fun channelRW(file: File): FileChannel {
        return RandomAccessFile(file, "rw").channel
    }

    private fun channelDRW(file: File): FileChannel {
        file.delete()
        return RandomAccessFile(file, "rw").channel
    }

    private fun listRecursive(file: File,
                              list: MutableList<FilePath>) {
        val files = file.listFiles()
        if (files != null) {
            for (child in files) {
                list.add(path(child))
                if (child.isDirectory) {
                    listRecursive(child, list)
                }
            }
        }
    }
}
